### PR TITLE
Migrate away from deprecated OpenTelemetry Tracer API

### PR DESCRIPTION
**Pull request by iNikem**

### CODE REVIEW

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/PathAttributeExtractor.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+
+import org.springframework.cloud.sleuth.http.HttpRequest;
+import org.springframework.cloud.sleuth.http.HttpResponse;
+import org.springframework.lang.Nullable;
+import org.springframework.util.StringUtils;
+
+class PathAttributeExtractor implements AttributesExtractor<HttpRequest, HttpResponse> {
+
+	private static final AttributeKey<String> HTTP_PATH = AttributeKey.stringKey("http.path");
```
Comment (mateuszrzeszutek) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/PathAttributeExtractor.java:31]: This constant seems to be unused

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/OtelHttpClientHandler.java [hunk 2]
```diff
@@ -39,156 +34,101 @@
...
-		String path = request.path();
-		if (path != null) {
-			span.setAttribute(SemanticAttributes.HTTP_ROUTE, path);
-		}
-		return context;
+		SpringHttpClientAttributesExtractor httpAttributesExtractor = new SpringHttpClientAttributesExtractor();
+		this.instrumenter = Instrumenter
+				.<HttpClientRequest, HttpClientResponse>newBuilder(openTelemetry, "org.springframework.cloud.sleuth",
+						HttpSpanNameExtractor.create(httpAttributesExtractor))
+				.addAttributesExtractor(new HttpRequestNetClientAttributesExtractor())
```
Comment (mateuszrzeszutek) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/OtelHttpClientHandler.java:70]: Should we set the status extractor too? (also in server handler)
```suggestion
				.setSpanStatusExtractor(HttpSpanStatusExtractor.create(httpAttributesExtractor))
				.addAttributesExtractor(new HttpRequestNetClientAttributesExtractor())
```

File: spring-cloud-sleuth-otel/pom.xml [hunk 3]
```diff
@@ -54,6 +54,10 @@
 			<groupId>io.opentelemetry</groupId>
 			<artifactId>opentelemetry-api</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>io.opentelemetry</groupId>
+			<artifactId>opentelemetry-api-metrics</artifactId>
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-otel/pom.xml:59]: Do we really need this?

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/HttpRequestNetClientAttributesExtractor.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2013-2020 the original author or authors.
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/HttpRequestNetClientAttributesExtractor.java:2]: 2021 (multiple occurrences)

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	@Nullable
+	@Override
+	protected String target(HttpServerRequest httpServerRequest) {
+		URI uri = toUri(httpServerRequest);
+		if (uri == null) {
+			return null;
+		}
+		return uri.getPath() + queryPart(uri);
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java:49]: As far as I remember the fragment is also the part of this according to the semconv specs.

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+	@Nullable
+	@Override
+	protected String route(HttpServerRequest httpServerRequest) {
+		return httpServerRequest.route();
+	}
+
+	@Nullable
+	@Override
+	protected String scheme(HttpServerRequest httpServerRequest) {
```
Comment (jonatan-ivanov) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java:70]: Wouldn't be `toUri(request).getScheme()` simpler?

File: spring-cloud-sleuth-otel/pom.xml (hunk 3 above)
Comment (iNikem) [reply spring-cloud-sleuth-otel/pom.xml:59]: `Instrumenter` class uses classes from Metrics API. But I will investigate why this explicit dependency was needed.

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java (hunk 6 above)
Comment (iNikem) [reply spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java:70]: I am generally quite reluctant to convert strings to URL/URI for simple cases. That conversion is quite expensive both CPU and memory wise.

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/PathAttributeExtractor.java (hunk 1 above)
Comment (iNikem) [reply spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/PathAttributeExtractor.java:31]: It is used now

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java (hunk 5 above)
Comment (iNikem) [reply spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java:49]: There is an open discussion currently about the fate of this attribute: https://github.com/open-telemetry/opentelemetry-specification/issues/2056

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/HttpRequestNetClientAttributesExtractor.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+	@Nullable
+	@Override
+	public String peerName(HttpRequest httpRequest, @Nullable HttpResponse httpResponse) {
+		return null;
+	}
+
+	@Override
+	public Integer peerPort(HttpRequest httpRequest, @Nullable HttpResponse httpResponse) {
+		return httpRequest == null ? null : httpRequest.remotePort();
```
Comment (mateuszrzeszutek) [inline spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/HttpRequestNetClientAttributesExtractor.java:46]: Is it possible for `httpRequest` to be null?

Comment (iNikem) [reply spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/HttpRequestNetClientAttributesExtractor.java:46]: Unfortunately, yes. At least in tests.

File: spring-cloud-sleuth-otel/pom.xml (hunk 3 above)
Comment (jonatan-ivanov) [reply spring-cloud-sleuth-otel/pom.xml:59]: Thanks, it is a little weird that we need this.

File: spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java (hunk 5 above)
Comment (jonatan-ivanov) [reply spring-cloud-sleuth-otel/src/main/java/org/springframework/cloud/sleuth/otel/bridge/SpringHttpServerAttributesExtractor.java:49]: 👍🏼 I would implement it in a way that confirms the current specs and change if the specs changed later.

**Pull request merged**